### DISCUSSION TITLE
Does 2D FillHoles has the concept of scipy's 'structure' ?

### DISCUSSION THREAD

**Discussion by ogencoglu** [DIPlib/diplib#173] [General] [answered]
In `scipy` `binary_fill_holes`, one can control the definition of "hole" with e.g. `structure=ndimage.generate_binary_structure(2, 2)`.

Does diplib has such control?

**Comment by crisluengo** [comment, marked as answer]
DIPlib’s [`FillHoles()`](https://diplib.org/diplib-docs/binary.html#dip-FillHoles-Image-CL-Image-L-dip-uint-) is controlled by a connectivity parameter. This is equivalent to choosing a 3x3 diamond or a 3x3 square structuring element in [`binary_fill_holes`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.binary_fill_holes.html). We don’t allow arbitrary neighborhoods because (1) you don’t necessarily get useful results with them, and (2) it makes the algorithm much more expensive.

SciPy’s function applies iterative dilations with the given structuring element. DIPlib’s function uses a queue-based algorithm that propagates a front. This is much more efficient than the repeated dilations.

If you want to ignore small gaps in this algorithm, you can apply a dilation before the hole filling, and a corresponding erosion after.

**Comment by ogencoglu** [comment]
Understood! Thanks for the swift reply!
